File: lib/models/Student/Student.go
package Student

import (
	"os"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/dynamodb"
)

type Student struct {
	Pk   string `json:"pk"`
	Name string `json:"name"`
}

var STUDENTS_TABLE string

var sess *session.Session
var ddb *dynamodb.DynamoDB

func Init(_sess **session.Session, _ddb **dynamodb.DynamoDB) {
	if *_sess == nil || *_ddb == nil {
		if os.Getenv("PROD") == "PROD" {
			*_sess = session.Must(session.NewSessionWithOptions(session.Options{SharedConfigState: session.SharedConfigEnable}))
			*_ddb = dynamodb.New(*_sess)

		} else if os.Getenv("TEST") == "TEST" {
			*_sess = session.Must(session.NewSessionWithOptions(session.Options{SharedConfigState: session.SharedConfigEnable}))
			*_ddb = dynamodb.New(*_sess, &aws.Config{
				Endpoint:   aws.String("http://localhost:8000"),
				DisableSSL: aws.Bool(true),
			})
		} else if os.Getenv("TEST_IN_DOCKER") == "TEST_IN_DOCKER" {
			*_sess = session.Must(session.NewSessionWithOptions(session.Options{SharedConfigState: session.SharedConfigEnable}))
			*_ddb = dynamodb.New(*_sess, &aws.Config{
				Endpoint:   aws.String("http://" + os.Getenv("DB_HOST") + ":8000"),
				DisableSSL: aws.Bool(true),
			})
		}
	}

	STUDENTS_TABLE = os.Getenv("STUDENTS_TABLE")

	sess = *_sess
	ddb = *_ddb
}

// TODO: add functions
// func CheckGivenPairs(Pair string) (*dynamodb.QueryOutput, error) {
// 	var checkPair *dynamodb.QueryInput = &dynamodb.QueryInput{
// 		TableName:              aws.String(STUDENTS_TABLE),
// 		KeyConditionExpression: aws.String("pk = :partitionKey and sk=:sortKey"),
// 		ExpressionAttributeValues: map[string]*dynamodb.AttributeValue{
// 			":partitionKey": {
// 				S: aws.String("pair"),
// 			},
// 			":sortKey": {
// 				S: aws.String(Pair),
// 			},
// 		},
// 	}
// 	return ddb.Query(checkPair)
// }

File: lib/models/Student/Student.js
const { DynamoDB } = require("aws-sdk");
const { Marshaller } = require("@aws/dynamodb-auto-marshaller");
const marshaller = new Marshaller({ unwrapNumbers: true });
const fs = require("fs");
const uuid4 = require("uuid4");

let options = {};
let STUDENTS_TABLE;

if (process.env.PROD) {
  options = {
    apiVersion: "2012-08-10",
    region: "eu-west-1",
  };
  const tableJSON = require("./dynamodb");
  STUDENTS_TABLE = tableJSON.tableName;
} else if (process.env.DEV || process.env.TEST) {
  options = {
    apiVersion: "2012-08-10",
    endpoint: "http://localhost:8000",
    region: "eu-west-1",
    sslEnabled: false,
  };
  const tableJSON = JSON.parse(
    fs.readFileSync(__dirname + "/dynamodb.dev.json", "utf8")
  );
  STUDENTS_TABLE = tableJSON.TableName;
}

const ddb = new DynamoDB(options);

// ! ========================================================

const Student = {
  async get(id) {
    const params = {
      TableName: STUDENTS_TABLE,
      KeyConditionExpression: "pk = :partitionKey",
      ExpressionAttributeValues: {
        ":partitionKey": { S: id },
      },
    };
    const result = await ddb.query(params).promise();
    if (result.Count === 0) {
      return undefined;
    }
    result.Items = result.Items.map((item) => {
      return marshaller.unmarshallItem(item);
    });
    return result.Items[0];
  },

  async getAll() {
    const params = {
      TableName: STUDENTS_TABLE,
    };
    const result = await ddb.scan(params).promise();
    result.Items = result.Items.map((item) => {
      return marshaller.unmarshallItem(item);
    });

    return result;
  },

  async create(name) {
    if (!name) {
      return "Missing parameters";
    }
    const item = {
      pk: uuid4(),
      name: name,
    };
    const params = {
      TableName: STUDENTS_TABLE,
      Item: marshaller.marshallItem(item),
    };
    await ddb.putItem(params).promise();
    return true;
  },

  async update(id, name) {
    if (!id || !name) {
      return "Missing parameters";
    }
    const items = {
      pk: id,
      name: name,
    };
    const params = {
      TableName: STUDENTS_TABLE,
      Item: marshaller.marshallItem(items),
    };
    await ddb.putItem(params).promise();
    return true;
  },

  async delete(id) {
    const params = {
      Key: {
        pk: {
          S: id,
        },
      },
      TableName: STUDENTS_TABLE,
    };
    await ddb.deleteItem(params).promise();
    return true;
  },
};

module.exports = Student;
